### PR TITLE
feat: implement Opslane OPSL.8 caching layer

### LINKED ISSUE #400: [FEAT] implement OPSL.8 caching layer for Opslane flagship
## Summary

Implement OPSL.8 Caching Layer for the Opslane flagship project.

## Affected Area

- `s11` flagship: `internal/cache/`, `internal/middleware/cache.go`
- Module spec: `modules/08-caching/README.md`

## What This Module Builds

- In-memory cache with bounded TTL and capacity (cache-aside pattern)
- `Cache` interface and `InMemoryStore` implementation with LRU-style eviction
- Tenant-scoped cache keys for order and payment reads
- Explicit invalidation on writes (order transitions, payment settlements)
- HTTP-level Cache-Control middleware for public endpoints
- Stampede prevention via singleflight on hot-key expiry
- Full behavioral test suite proving TTL expiry, invalidation, and stampede safety

## Required Files

- `internal/cache/cache.go` — Cache interface, TTL entry, config
- `internal/cache/store.go` — InMemoryStore with sync.RWMutex + singleflight
- `internal/cache/cache_test.go` — Behavioral tests
- `internal/middleware/cache.go` — HTTP Cache-Control header middleware

## Proof Surface

- `go test ./11-flagship/01-opslane/internal/cache/...` passes
- `go run ./11-flagship/01-opslane/scripts/progress.go` shows OPSL.8 complete

### PR DESCRIPTION
**Pull request by rasel9t6**
## Summary

Implements OPSL.8 Caching Layer for the Opslane flagship project.

Closes #400

### New Files
- `internal/cache/cache.go` — Cache interface, key helpers, Invalidator, Singleflight, NoopCache
- `internal/cache/store.go` — InMemoryStore with bounded capacity, TTL, and background janitor
- `internal/cache/cache_test.go` — 15 behavioral tests
- `internal/middleware/cache.go` — HTTP Cache-Control and NoCache middleware
- `modules/08-caching/SURFACE.md` — Implementation map

### Updated Files
- `modules/08-caching/README.md` — Full spec with mermaid diagrams, proof surface, machine view
- `MODULES.md` — OPSL.8 marked as implemented with proof commands
- `README.md` — Module 8 snapshot, OPSL.9 next step

## Design Decisions

1. **PostgreSQL stays system of record** — cache is additive, never authoritative
2. **Insert-order eviction** — simpler than true LRU, acceptable for teaching codebase
3. **Copy-on-read/write** — prevents callers from mutating cached data
4. **Singleflight** — prevents thundering herd on hot key expiry
5. **Lazy expiry + janitor** — expired entries removed on read and periodically swept
6. **Invalidation after writes** — never before; if DB write fails, cache keeps correct old value

## Proof Surface

```
go test -count=1 ./11-flagship/01-opslane/internal/cache/...
ok  	github.com/rasel9t6/the-go-engineer/.../internal/cache	0.300s

go run ./11-flagship/01-opslane/scripts/progress.go
OPSL.8   Caching Layer                        complete
```

## CI Checks
- [x] `go build ./...`
- [x] `go vet ./...`
- [x] `go test ./...` (all 11 packages pass)
- [x] `go fmt ./...` (no diff)
- [x] `go run ./scripts/validate_curriculum.go` (511 files validated)

### CODE REVIEW

File: 11-flagship/01-opslane/internal/cache/cache.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+		return c.val, c.err
+	}
+
+	c := &call{}
+	c.wg.Add(1)
+	sf.calls[key] = c
+	sf.mu.Unlock()
+
+	c.val, c.err = fn()
+	c.wg.Done()
```
[the remark this answers is not in the document]
Comment (rasel9t6) [reply 11-flagship/01-opslane/internal/cache/cache.go]: Fixed in b0000a9: added deferred cleanup and TestSingleflightPanicRecovery.

File: 11-flagship/01-opslane/internal/cache/cache.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+	c.wg.Add(1)
+	sf.calls[key] = c
+	sf.mu.Unlock()
+
+	// Ensure cleanup runs even if fn() panics
+	defer func() {
+		sf.mu.Lock()
+		delete(sf.calls, key)
+		sf.mu.Unlock()
+		c.wg.Done()
```
[the remark this answers is not in the document]
Comment (rasel9t6) [reply 11-flagship/01-opslane/internal/cache/cache.go]: Fixed in ab2fbd1: c.wg.Done() is now called before deleting the active singleflight call.

**Comment by rasel9t6** [comment]
## Final readiness check

### Changes
- Bounded in-memory cache with TTL, insert-order eviction, and background janitor
- Copy-on-read/write for mutation safety
- Singleflight stampede prevention with panic-safe error propagation
- Invalidator for write-through cache invalidation
- HTTP Cache-Control and NoCache middleware
- NoopCache for disabled-cache environments
- 16 behavioral tests covering edge cases

### Review findings handled
- **P2 (Codex)**: Singleflight panic propagation — fixed. Leader and waiters now receive a concrete error instead of nil/nil when fn() panics. Uses named returns and recover() in defer.
- Test updated to verify error message content and second-call recovery.

### Validation
- `go test -v ./11-flagship/01-opslane/internal/cache/...` — 16/16 PASS
- `go build ./...` — passes
- `go vet ./...` — passes
- `go run ./scripts/validate_curriculum.go` — passes (508 files, 12 sections, 215 items)
- CI on PR — SUCCESS

### Remaining risks
- Race detector not runnable locally on Windows (CGO required). CI runs it on Ubuntu.
- removeFromOrder is O(n) — acceptable for bounded teaching cache.
